render/reply/formatter_test.go

```diff
@@ -0,0 +1,240 @@
+package reply
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/lomik/graphite-clickhouse/finder"
+	"github.com/lomik/graphite-clickhouse/helper/client"
+	"github.com/lomik/graphite-clickhouse/helper/point"
+	"github.com/lomik/graphite-clickhouse/pkg/alias"
+	"github.com/lomik/graphite-clickhouse/render/data"
+)
+
+var results = []client.Metric{
+	{
+		Name:           "test.metric1",
+		PathExpression: "test.*",
+		StartTime:      1688990040,
+		StepTime:       60,
+		StopTime:       1688990520,
+		Values: func() []float64 {
+			temp := emptyValues(8)
+			temp[2] = 3
+			return temp
+		}(),
+	},
+	{
+		Name:           "test.metric2",
+		PathExpression: "test.*",
+		StartTime:      1688990040,
+		StepTime:       60,
+		StopTime:       1688990520,
+		Values:         emptyValues(8),
+	},
+	{
+		Name:           "test.metric3",
+		PathExpression: "test.*",
+		StartTime:      1688990040,
+		StepTime:       60,
+		StopTime:       1688990520,
+		Values:         emptyValues(8),
+	},
+}
+
+func TestFormatterReply(t *testing.T) {
+	formatters := []struct {
+		impl   Formatter
+		name   string
+		format client.FormatType
+	}{
+		{&V3PB{}, "v3pb", client.FormatPb_v3},
+		{&V2PB{}, "v2pb", client.FormatPb_v2},
+		{&JSON{}, "json", client.FormatJSON},
+		{&Pickle{}, "pickle", client.FormatPickle},
+	}
+	tests := []struct {
+		name  string
+		input data.CHResponses
+		// result when CHResponse.AppendOutEmptySeries is false
+		expected0 []client.Metric
+		// result when CHResponse.AppendOutEmptySeries is true
+		expected1 []client.Metric
```

Comment:
may be better expectedWithoutEmpty and expectedWithEmpty instead of magic names expected0 / expected1?

Reply:
I tried using names like this, but they sounded ambiguous too. In this case,
`expected0 == expectedWithEmpty` ?

Reply:
expected0 => expectedWithoutEmpty ?